invokeai/app/services/model_install/model_install_default.py

```diff
@@ -1,60 +1,69 @@
 """Model installation class."""
 
+import re
 import threading
+import time
 from hashlib import sha256
-from logging import Logger
 from pathlib import Path
-from queue import Queue
+from queue import Empty, Queue
 from random import randbytes
 from shutil import copyfile, copytree, move, rmtree
+from tempfile import mkdtemp
 from typing import Any, Dict, List, Optional, Set, Union
 
+from huggingface_hub import HfFolder
+from requests import Session
+
 from invokeai.app.services.config import InvokeAIAppConfig
-from invokeai.app.services.events import EventServiceBase
-from invokeai.app.services.model_records import DuplicateModelException, ModelRecordServiceBase, UnknownModelException
+from invokeai.app.services.download import DownloadJob, DownloadQueueServiceBase
+from invokeai.app.services.events.events_base import EventServiceBase
+from invokeai.app.services.model_records import DuplicateModelException, ModelRecordServiceBase, ModelRecordServiceSQL
 from invokeai.backend.model_manager.config import (
     AnyModelConfig,
     BaseModelType,
     InvalidModelConfigException,
+    ModelRepoVariant,
     ModelType,
 )
 from invokeai.backend.model_manager.hash import FastModelHash
+from invokeai.backend.model_manager.metadata import (
+    AnyModelRepoMetadata,
+    CivitaiMetadataFetch,
+    HuggingFaceMetadataFetch,
+    ModelMetadataStore,
+    ModelMetadataWithFiles,
+    RemoteModelFile,
+)
 from invokeai.backend.model_manager.probe import ModelProbe
 from invokeai.backend.model_manager.search import ModelSearch
 from invokeai.backend.util import Chdir, InvokeAILogger
+from invokeai.backend.util.devices import choose_precision, choose_torch_device
 
 from .model_install_base import (
+    CivitaiModelSource,
+    HFModelSource,
     InstallStatus,
     LocalModelSource,
     ModelInstallJob,
     ModelInstallServiceBase,
     ModelSource,
+    URLModelSource,
 )
 
-# marker that the queue is done and that thread should exit
-STOP_JOB = ModelInstallJob(
-    source=LocalModelSource(path="stop"),
-    local_path=Path("/dev/null"),
-)
+TMPDIR_PREFIX = "tmpinstall_"
```

Comment:
`tempfile` works great as a context manager, but I couldn't figure out how to create a singular context for the multithreaded downloads. The docs say that `tempfile` will also be deleted at garbage collection time, but I couldn't get it to work reliably. So I ended up managing the thing manually.

Reply:
Gotcha. Wonder if @RyanJDick has any ideas.

Reply:
My best guess for why it wasn't working correctly is because we are using daemon threads, and may not be shutting them down cleanly.

From the [docs](https://docs.python.org/3/library/threading.html):
> Daemon threads are abruptly stopped at shutdown. Their resources (such as open files, database transactions, etc.) may not be released properly. 

Does that match your observations @lstein?

Those dangling temp files _should_ be stored in the OS's temp directory, so would be cleaned up by the OS on the next reboot. But, it's probably not a great idea to rely on that.

My proposal is that we add a `TODO` comment explaining why we handle this manually, including our best hypothesis for why `tempfile` wasn't working.

In the future, I think all of the multi-threading stuff can be radically cleaned up and this problem might just go away.

Reply:
I've added the TODO comment. 

I would really like to have these tmp directories cleaned up after the download and install is completed rather than waiting for the server to exit or the whole system to be rebooted. They eat up a lot of disk space. The current manual system does remove these directories at the end of the download, but they may dangle if the server is interrupted, in which case they are cleaned up the next time the server is started.